Complete Flutter Testing Guide

ALIRAZA17 opened the pull request:
<!--
Thanks for submitting a pull request! Please provide enough information so that others can review your pull request.
-->

**What kind of change does this PR introduce?**

I have written a complete Flutter Testing Guide Document which explains everything from adding required packages to creating a test file and then running the test cases.

**Issue Number:**

Fixes #597 

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-docs/blob/develop/CONTRIBUTING.md)?**

Yes

ALIRAZA17 committed:
Added Complete Flutter Testing Guide

ALIRAZA17 committed:
Added Image in Document

ALIRAZA17 committed:
Update flutter_testing_guide.md

ALIRAZA17 committed:
Update flutter_testing_guide.md

ALIRAZA17 commented:
@palisadoes @literalEval @noman2002 I have added the Flutter Testing Guide. Can you please review it and if there's anything missing I can add?

literalEval commented:
Awesome work @ALIRAZA17 :) 

I have two suggestions.

1. Add a better image.
2. Briefly elaborate on important parts of the testing system that Talawa has. The idea is that there are already a lot of tutorial on flutter testing. If we are writing one for ourselves, it should be customized to our needs. You can add things like Mocks for such and such classes already exist but all of the methods/attributes are not mocked for now etc.

literalEval commented:
Please note that you should only follow these suggestions if @palisadoes agrees with them. Otherwise you are good to go.

palisadoes commented:
@ALIRAZA17 I have done two Flutter courses and never written an app outside of it. Take the advice of @literalEval for now and let me know when you are done.

ALIRAZA17 commented:
> @ALIRAZA17 I have done two Flutter courses and never written an app outside of it. Take the advice of @literalEval for now and let me know when you are done.

ok ok Im on it. I'll let you know!!

ALIRAZA17 commented:
> Awesome work @ALIRAZA17 :)
> 
> I have two suggestions.
> 
> 1. Add a better image.
> 2. Briefly elaborate on important parts of the testing system that Talawa has. The idea is that there are already a lot of tutorial on flutter testing. If we are writing one for ourselves, it should be customized to our needs. You can add things like Mocks for such and such classes already exist but all of the methods/attributes are not mocked for now etc.

on it!

ALIRAZA17 committed:
Update flutter_testing_guide.md

ALIRAZA17 committed:
Add files via upload

ALIRAZA17 committed:
Updated Image

ALIRAZA17 commented:
@palisadoes @literalEval Can you please review it again I have updated the image and also added an example of test from talawa app. Is there anything I should remove or add in this doc?

literalEval commented:
@ALIRAZA17 please remove the earlier image which you are not using now.

ALIRAZA17 committed:
Delete testing_guide.png

ALIRAZA17 commented:
> @ALIRAZA17 please remove the earlier image which you are not using now.

@literalEval done

literalEval commented:
@palisadoes looks good to me. Just add the file in `sidebar.js`

Cioppolo14 reviewed (changes requested):
Please edit sidebars.js to include this new page.

ALIRAZA17 commented:
> Please edit sidebars.js to include this new page. 

@Cioppolo14 is there something I have to edit?

literalEval commented:
@ALIRAZA17 open `sidebar.js` and add your filename in there. Open `sidebar.js` and you'll understand what to do.

ALIRAZA17 commented:
> @ALIRAZA17 open `sidebar.js` and add your filename in there. Open `sidebar.js` and you'll understand what to do.

Ok ok!!

ALIRAZA17 committed:
Update sidebars.js

ALIRAZA17 commented:
> @ALIRAZA17 open `sidebar.js` and add your filename in there. Open `sidebar.js` and you'll understand what to do.

@literalEval Added it can you check it if its ok?

ALIRAZA17 commented:
> > @ALIRAZA17 open `sidebar.js` and add your filename in there. Open `sidebar.js` and you'll understand what to do.
> 
> @literalEval @Cioppolo14 Added it can you check it if its ok?

ALIRAZA17 commented:
> > > @ALIRAZA17 open `sidebar.js` and add your filename in there. Open `sidebar.js` and you'll understand what to do.
> > 
> > 
> > @literalEval @Cioppolo14 @palisadoes Added it can you check it if its ok?

literalEval commented:
Did it show up properly in the sidebar of your local build ?

ALIRAZA17 committed:
Merge branch 'PalisadoesFoundation:develop' into MuhammadAliRazaDar/Flutter_Testing_Guide

ALIRAZA17 committed:
Updated sidebar.js

ALIRAZA17 commented:
> Did it show up properly in the sidebar of your local build ?

@literalEval Now it is running There was a mistake from my side. Its good to go now

palisadoes reviewed (changes requested):
Please update the PR. 

1. The flutter testing guide is not visible in the sidebar. You need to correctly update sidebar.js to add the entry
2. The markdown file you added does not have the correct entries at the top like the other markdown files in the directory tree to be recognized by sidebar.js
3. Remove the image and references to it. It is not necessary.

ALIRAZA17 commented:
> Please update the PR.
> 
> 1. The flutter testing guide is not visible in the sidebar. You need to correctly update sidebar.js to add the entry
> 2. The markdown file you added does not have the correct entries at the top like the other markdown files in the directory tree to be recognized by sidebar.js
> 3. Remove the image and references to it. It is not necessary.

@palisadoes Im facing a problem of document id because my document id doesnt exist. Where to add that id of my document to configure it??

palisadoes commented:
Look at the other files in the directory. You will see their IDs at the top

ALIRAZA17 commented:
![image](https://user-images.githubusercontent.com/75278584/229314241-c9e5e6b9-3459-433e-968f-758d0bfb992d.png)
When I add this in my document I face this error
![image](https://user-images.githubusercontent.com/75278584/229314257-ec43e2ea-21f3-43a6-b414-803ab3b40c64.png)

ALIRAZA17 commented:
> ![image](https://user-images.githubusercontent.com/75278584/229314241-c9e5e6b9-3459-433e-968f-758d0bfb992d.png) 

When I add this in my document I face this error 

![image](https://user-images.githubusercontent.com/75278584/229314257-ec43e2ea-21f3-43a6-b414-803ab3b40c64.png)

ALIRAZA17 commented:
> Look at the other files in the directory. You will see their IDs at the top

Ive added it but it gave me that error I just showed you

palisadoes commented:
Side bar does not check the file name
It checked the IDs of the files in the directory

ALIRAZA17 commented:
> Side bar does not check the file name It checked the IDs of the files in the directory

Ive updated the file and added the id but it gives me an error of invalid document id. If I remove that portion it works fine on my local machine and my document appears on the web page

palisadoes commented:
Please follow the convention we have already. All files have IDs and the ID is referenced in the sidebar.

The first part of the entry in the sidebar is the directory tree and the last item of the entry is the Id.

Look at the other files in the other directories as a guide.

We need to be consistent or else the risk of breaking out documentation gets too high

ALIRAZA17 committed:
Updated SiderBar.js

ALIRAZA17 commented:
> Please follow the convention we have already. All files have IDs and the ID is referenced in the sidebar.
> 
> The first part of the entry in the sidebar is the directory tree and the last item of the entry is the Id.
> 
> Look at the other files in the other directories as a guide.
> 
> We need to be consistent or else the risk of breaking out documentation gets too high

@palisadoes Can you check it now the error is resolved

palisadoes commented:
Fix this:

![image](https://user-images.githubusercontent.com/16875803/229328305-097e02f7-97a7-42e7-bd0a-cc6300658ff1.png)

palisadoes commented on docs/developers/talawa/flutter_testing_guide.md:
```diff
@@ -0,0 +1,271 @@
+
+4- The `when` function is used to set up a mock behavior for the `gqlAuthMutation` method of the `_dbFunctions` object. The mocked behavior returns a `QueryResult` object that simulates the result of a GraphQL mutation. The `QueryResult` object contains a map with a key of 'sendMessageToDirectChat', which contains a value that represents the returned ChatMessage object from the mutation.
+
+Overall, this test verifies that the sendMessageToDirectChat method correctly triggers a GraphQL mutation and correctly handles the returned data by updating the _chatMessageController object with the expected ChatMessage object.
```
_chatMessageController should be code

ALIRAZA17 replied:
![image](https://user-images.githubusercontent.com/75278584/229331407-a999b227-c6e4-4a6f-b158-241df2180e8a.png)

palisadoes commented on docs/developers/talawa/flutter_testing_guide.md:
```diff
@@ -0,0 +1,271 @@
+
+3- Finally, the test verifies that the `_chatMessageController` object has been updated with the correct `ChatMessage` object that was received from the mocked `GraphQL mutation` result.
+
+4- The `when` function is used to set up a mock behavior for the `gqlAuthMutation` method of the `_dbFunctions` object. The mocked behavior returns a `QueryResult` object that simulates the result of a GraphQL mutation. The `QueryResult` object contains a map with a key of 'sendMessageToDirectChat', which contains a value that represents the returned ChatMessage object from the mutation.
```
`sendMessageToDirectChat`

palisadoes commented on docs/developers/talawa/flutter_testing_guide.md:
````diff
@@ -0,0 +1,271 @@
+var myMock = MockMyObject();
+when(myMock.myMethod()).thenReturn('my result');
+```
+This will replace the `myMethod()` method on the `myMock` object with a stub that always returns the string 'my result'.
````
`my result` should be code.

ALIRAZA17 replied:
![image](https://user-images.githubusercontent.com/75278584/229331426-09dbf74c-01cf-4fac-8059-97e88f6a6010.png)

palisadoes commented on docs/developers/talawa/flutter_testing_guide.md:
````diff
@@ -0,0 +1,271 @@
+```dart
+when(myMock.myMethod(any)).thenReturn('my result');
+```
+This will stub the `myMethod()` method to always return 'my result', regardless of the input value.
````
`my result` should be code

ALIRAZA17 replied:
![image](https://user-images.githubusercontent.com/75278584/229331452-07f0f1d5-25f7-458f-a316-f09f7b1a44cd.png)

palisadoes commented on docs/developers/talawa/flutter_testing_guide.md:
````diff
@@ -0,0 +1,271 @@
+    });
+    ```
+    
+   ### Example
````
This needs to be against the margin

ALIRAZA17 replied:
I couldn't get it?

palisadoes commented on docs/developers/talawa/flutter_testing_guide.md:
```diff
@@ -0,0 +1,271 @@
+
+// Create the mock object in your test case
+final calculator = MockCalculator();
+
```
Fix this:

![image](https://user-images.githubusercontent.com/16875803/229328507-5baf5dc5-7d99-4add-863a-4372bb5316fc.png)

palisadoes commented on docs/developers/talawa/flutter_testing_guide.md:
```diff
@@ -0,0 +1,271 @@
+
+Here is a breakdown of what this test does
+
+1- The test starts by defining a mock object for the `_dbFunctions` class using the when function from the `Mockito` package. The mock object is set up to return a `QueryResult` object that simulates the result of a GraphQL mutation when the `gqlAuthMutation` method is called with the correct query and variables.
```
use:

1.
2.
3.
4.

The CSS will handle the rest

palisadoes commented on docs/developers/talawa/flutter_testing_guide.md:
```diff
@@ -0,0 +1,271 @@
+
+# Flutter Testing Guide
+
+## Introduction
```
Remove heading

ALIRAZA17 replied:
which one

palisadoes replied:
Flutter Testing Guide

ALIRAZA17 committed:
Updated the document

palisadoes commented on docs/developers/talawa/flutter_testing_guide.md:
```diff
@@ -0,0 +1,271 @@
+
+## Why do we need tests?
+
+Tests ensure that the code you write works as intended. They help to catch bugs early in the development process, which saves time and effort in the long run. Writing tests also makes the code more reliable and maintainable. 
```
Merge this section into that above. It is repetitive

ALIRAZA17 commented:
> Fix this:
> 
> ![image](https://user-images.githubusercontent.com/16875803/229328305-097e02f7-97a7-42e7-bd0a-cc6300658ff1.png)

@palisadoes Done!!

palisadoes commented on docs/developers/talawa/flutter_testing_guide.md:
```diff
@@ -0,0 +1,271 @@
+It's also a good idea to use boundary testing, where you test the boundaries between different input values. For example, if your function takes an input between 0 and 100, you should test it with values of 0, 1, 99, 100, and values just above and below these boundaries.
+
+
+## How to start with Flutter Testing?
```
Change to "Getting Started"

palisadoes commented on docs/developers/talawa/flutter_testing_guide.md:
```diff
@@ -0,0 +1,271 @@
+
+## How to start with Flutter Testing?
+
+1. Add the `flutter_test` package to your `pubspec.yaml` file.
```
Add an introductory sentence before bullets. It makes the flow better

palisadoes commented on docs/developers/talawa/flutter_testing_guide.md:
```diff
@@ -0,0 +1,271 @@
+ 
+4. Run your tests using the flutter test command. This will run all the tests in your project's test directory.
+
+### Basic Example
```
Change to "Basic Test Example", otherwise the question is Example of what? The titles need to be self explanatory.

palisadoes commented on docs/developers/talawa/flutter_testing_guide.md:
```diff
@@ -0,0 +1,271 @@
+The cost of removing defects increases exponentially. A defect caught in requirement and design phase costs less to fix than an error caught in the software maintenance cycle.
+
+
+## How to cover all the corner cases?
```
Change to "Understanding Corner Cases"

palisadoes commented on docs/developers/talawa/flutter_testing_guide.md:
```diff
@@ -0,0 +1,271 @@
+
+
+## How to cover all the corner cases?
+
```
Add a brief sentence about what a corner case is. Don't expect everyone to know.

palisadoes commented on docs/developers/talawa/flutter_testing_guide.md:
```diff
@@ -0,0 +1,271 @@
+If the test fails (for example, if the add method in calculator.dart was implemented incorrectly), you will see an error message in your terminal indicating what went wrong.
+
+
+## What are mocks and how to generate them?
```
1. Create  a title "In Depth Example"
2. Add an introductory sentence
3. Make the mocks, stubs and test sections fall under it

ALIRAZA17 replied:
I should create this heading under the heading of what are mocks and how to generate them?

palisadoes commented on docs/developers/talawa/flutter_testing_guide.md:
```diff
@@ -0,0 +1,271 @@
+If the test fails (for example, if the add method in calculator.dart was implemented incorrectly), you will see an error message in your terminal indicating what went wrong.
+
+
+## What are mocks and how to generate them?
```
Change to just "Mocks". Some of the titles are so long that they wrap twice in the right hand side bar.

palisadoes commented on docs/developers/talawa/flutter_testing_guide.md:
```diff
@@ -0,0 +1,271 @@
+  arguments.
+
+
+## How to Stub Methods and Classes?
```
Change to "Stubs" Add another #

palisadoes commented on docs/developers/talawa/flutter_testing_guide.md:
````diff
@@ -0,0 +1,271 @@
+```
+This will stub the `myMethod()` method to always return 'my result', regardless of the input value.
+
+## Talawa App Tests Example on How to use Mocks and Stubs Correctly
````
Change to "Example Test  of Mocks and Stubs" add #

ALIRAZA17 replied:
'#' should be single or added into the previous '##'?

palisadoes commented on docs/developers/talawa/flutter_testing_guide.md:
```diff
@@ -0,0 +1,271 @@
+
+There is a function in talawa which provides chat service in the app. The file is located in `talawa/lib/services/chat_service.dart`. Unit Tests for this file have been written to test all the methods written in the file for instance `sendMessageToDirectChat` and `getDirectChatsByUserId` methods. These tests are written in the file `talawa\test\service_tests\chat_service_test.dart`. These tests h ...
+
+### sendMessageToDirectChat Method
```
1. Add # 
2. The title doesn't explain what you are trying to do.
3. Add some introductory text below the title. It makes the flow better. You have a title and then code.
4. If this is the code to be tested then say so. It's not clear.

ALIRAZA17 replied:
@palisadoes Im a bit confused here. Do I need to add a single '#' here?

palisadoes commented on docs/developers/talawa/flutter_testing_guide.md:
````diff
@@ -0,0 +1,271 @@
+  }
+```
+
+Test written for this method looks like this
````
There should be a title here like "Sample Mock and Test Code" with the text below.

palisadoes commented on docs/developers/talawa/flutter_testing_guide.md:
````diff
@@ -0,0 +1,271 @@
+    })
+```
+
+Here is a breakdown of what this test does
````
There should be a title here too. The segmentation makes the readability better.

ALIRAZA17 replied:
What should be the title here?

palisadoes replied:
Think of something appropriate given my prior advice

palisadoes commented on docs/developers/talawa/flutter_testing_guide.md:
```diff
@@ -0,0 +1,271 @@
+Overall, this test verifies that the sendMessageToDirectChat method correctly triggers a GraphQL mutation and correctly handles the returned data by updating the _chatMessageController object with the expected ChatMessage object.
+
+
+## What to do if you find a bug while writing a test for that file?
```
Change this to "Toubleshooting"

palisadoes reviewed (changes requested):
1. Please make the suggested changes. 
2. The flow will be much better with them.
3. Please make sure the menuing on the right hand side makes sense when you are done.

palisadoes commented on docs/developers/talawa/flutter_testing_guide.md:
````diff
@@ -0,0 +1,268 @@
+ 4. Use the mock object in your test cases:
+
+    ```dart
+    test('my test case', () {
````
Fix

![image](https://user-images.githubusercontent.com/16875803/229329287-5873806c-f651-427a-b1ee-9b2a361956ee.png)

ALIRAZA17 committed:
Updated the Document

ALIRAZA17 commented:
@palisadoes I have commited again with all the changes you requested. I was confused about the used of '#' on certain heading so if you review them again to see if there are any mistakes left?

ALIRAZA17 committed:
Corrected Margins

palisadoes commented on docs/developers/talawa/flutter_testing_guide.md:
```diff
@@ -0,0 +1,274 @@
+ 
+4. Run your tests using the flutter test command. This will run all the tests in your project's test directory.
+
+### Basic Test Example
```
Remove a #

palisadoes commented on docs/developers/talawa/flutter_testing_guide.md:
````diff
@@ -0,0 +1,274 @@
+```
+If the test fails (for example, if the add method in calculator.dart was implemented incorrectly), you will see an error message in your terminal indicating what went wrong.
+
+### In Depth Example
````
1. Remove a #
2. Make it Examples

palisadoes commented on docs/developers/talawa/flutter_testing_guide.md:
````diff
@@ -0,0 +1,274 @@
+  ```
+
+
+### Example
````
#### Mocks Test Example

palisadoes commented on docs/developers/talawa/flutter_testing_guide.md:
````diff
@@ -0,0 +1,274 @@
+```
+This will stub the `myMethod()` method to always return `my result`, regardless of the input value.
+
+# Example Test of Mocks and Stubs
````
#### Mocks and Stubs Test Example

palisadoes commented on docs/developers/talawa/flutter_testing_guide.md:
```diff
@@ -0,0 +1,274 @@
+
+There is a function in talawa which provides chat service in the app. The file is located in `talawa/lib/services/chat_service.dart`. Unit Tests for this file have been written to test all the methods written in the file for instance `sendMessageToDirectChat` and `getDirectChatsByUserId` methods. These tests are written in the file `talawa\test\service_tests\chat_service_test.dart`. These tests h ...
+
+# sendMessageToDirectChat Method
```
##### "Method Under Test"

palisadoes commented on docs/developers/talawa/flutter_testing_guide.md:
````diff
@@ -0,0 +1,274 @@
+  }
+```
+
+# Sample Mock and Test Code
````
Change to ##### Sample Mock and Test Code

palisadoes commented on docs/developers/talawa/flutter_testing_guide.md:
````diff
@@ -0,0 +1,274 @@
+    })
+```
+
+# Test Explanation
````
Change to ##### Test Explanation

palisadoes commented on docs/developers/talawa/flutter_testing_guide.md:
```diff
@@ -0,0 +1,274 @@
+There is a function in talawa which provides chat service in the app. The file is located in `talawa/lib/services/chat_service.dart`. Unit Tests for this file have been written to test all the methods written in the file for instance `sendMessageToDirectChat` and `getDirectChatsByUserId` methods. These tests are written in the file `talawa\test\service_tests\chat_service_test.dart`. These tests h ...
+
+# sendMessageToDirectChat Method
+
```
Add introductory text that states that in this example we'll be testing the `sendMessageToDirectChat` method. It makes the flow better.

palisadoes reviewed (changes requested):
Please make the requested changes.

ALIRAZA17 committed:
Updated The doc with relevat changes

ALIRAZA17 commented:
@palisadoes Done!!

palisadoes commented on docs/developers/talawa/flutter_testing_guide.md:
````diff
@@ -0,0 +1,274 @@
+```
+This will stub the `myMethod()` method to always return `my result`, regardless of the input value.
+
+# Mocks and Stubs Test Example
````
This should be:

`### Mocks and Stubs Test Example`

The title doesn't make sense as a main heading if it's under Stubs.

palisadoes commented on docs/developers/talawa/flutter_testing_guide.md:
```diff
@@ -0,0 +1,274 @@
+
+In this example, we'll be testing the `sendMessageToDirectChat` method from our application. This method is responsible for sending direct messages between two users in a private chat. The `sendMessageToDirectChat` method is critical to the functionality of our application, and we need to ensure that it works correctly under a variety of conditions. To do so, we'll be using a combination of manua ...
+
+# Method Under Test
```
This should be:

`#### Method Under Test`

The title doesn't make sense as a main heading if it's under Stubs.

palisadoes commented on docs/developers/talawa/flutter_testing_guide.md:
````diff
@@ -0,0 +1,274 @@
+  }
+```
+
+##### Sample Mock and Test Code
````
This should be:

`#### Sample Mock and Test Code`

palisadoes commented on docs/developers/talawa/flutter_testing_guide.md:
````diff
@@ -0,0 +1,274 @@
+    })
+```
+
+##### Test Explanation
````
`#### Test Explanation`

palisadoes reviewed (changes requested):
That should be it.

ALIRAZA17 committed:
Added Requested Changes

ALIRAZA17 commented:
@palisadoes Done!!

literalEval commented:
@ALIRAZA17 please have a closer look at the guide in the coming time and any feedback it gets (especially from newcomers). If anyone suggests any additions, please think about them carefully. If a beginner learns something from your guide, then only your work for it will be successful :)

Suggesting this as a fellow contributor :)

ALIRAZA17 commented:
> @ALIRAZA17 please have a closer look at the guide in the coming time and any feedback it gets (especially from newcomers). If anyone suggests any additions, please think about them carefully. If a beginner learns something from your guide, then only your work for it will be successful :)
> 
> Suggesting this as a fellow contributor :)

Indeed I will keep an eye on this. Thanks for the input!!
